examples/tpm2-seal-unseal/main.go

```diff
@@ -0,0 +1,226 @@
+)
+
+func main() {
+	flag.Parse()
+
+	if *pcr < 0 || *pcr > 23 {
+		fmt.Fprintf(os.Stderr, "Invalid flag 'pcr': value %d is out of range", *pcr)
+		os.Exit(1)
+	}
+
+	err := run(*pcr, *tpmPath)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "Error: %v\n", err)
+		os.Exit(1)
+	}
+}
+
+func run(pcr int, tpmPath string) (retErr error) {
+	// Open the TPM
+	rwc, err := tpm2.OpenTPM(tpmPath)
+	if err != nil {
+		return fmt.Errorf("can't open TPM %q: %v", tpmPath, err)
+	}
+	defer func() {
+		if err := rwc.Close(); err != nil {
+			retErr = fmt.Errorf("%v\ncan't close TPM %q: %v", retErr, tpmPath, err)
+		}
+	}()
+
+	// Create the parent key against which to seal the data
+	srkPassword := ""
+	srkHandle, _, err := tpm2.CreatePrimary(rwc, tpm2.HandleOwner, tpm2.PCRSelection{}, "", srkPassword, srkTemplate)
+	if err != nil {
+		return fmt.Errorf("can't create primary key: %v", err)
+	}
+	defer func() {
+		if err := tpm2.FlushContext(rwc, srkHandle); err != nil {
+			retErr = fmt.Errorf("%v\nunable to flush SRK handle %q: %v", retErr, srkHandle, err)
+		}
+	}()
+	fmt.Printf("Created parent key with handle: 0x%x\n", srkHandle)
+
+	// Note the value of the pcr against which we will seal the data
+	pcrVal, err := tpm2.ReadPCR(rwc, pcr, tpm2.AlgSHA256)
+	if err != nil {
+		return fmt.Errorf("unable to read PCR: %v", err)
+	}
+	fmt.Printf("PCR %v value: 0x%x\n", pcr, pcrVal)
+
+	// Get the authorization policy that will protect the data to be sealed
+	objectPassword := "objectPassword"
+	sessHandle, policy, err := policyPCRPasswordSession(rwc, pcr, objectPassword)
+	if err != nil {
+		return fmt.Errorf("unable to get policy: %v", err)
+	}
+	if err := tpm2.FlushContext(rwc, sessHandle); err != nil {
+		return fmt.Errorf("unable to flush session: %v", err)
+	}
+	fmt.Printf("Created authorization policy: 0x%x\n", policy)
+
+	// Seal the data to the parent key and the policy
+	dataToSeal := []byte("secret")
+	fmt.Printf("Data to be sealed: 0x%x\n", dataToSeal)
+	privateArea, publicArea, err := tpm2.Seal(rwc, srkHandle, srkPassword, objectPassword, policy, dataToSeal)
+	if err != nil {
+		return fmt.Errorf("unable to seal data: %v", err)
+	}
+	fmt.Printf("Sealed data: 0x%x\n", privateArea)
+
+	// Load the sealed data into the TPM.
+	objectHandle, _, err := tpm2.Load(rwc, srkHandle, srkPassword, publicArea, privateArea)
+	if err != nil {
+		return fmt.Errorf("unable to load data: %v", err)
+	}
+	defer func() {
+		if err := tpm2.FlushContext(rwc, objectHandle); err != nil {
+			retErr = fmt.Errorf("%v\nunable to flush object handle %q: %v", retErr, objectHandle, err)
+		}
+	}()
+	fmt.Printf("Loaded sealed data with handle: 0x%x\n", objectHandle)
+
+	// Unseal the data
+	unsealedData, err := unseal(rwc, pcr, objectPassword, objectHandle)
+	if err != nil {
+		return fmt.Errorf("unable to unseal data: %v", err)
+	}
+	fmt.Printf("Unsealed data: 0x%x\n", unsealedData)
+
+	// Try to unseal the data with the wrong password
```

Comment:
I still think that all of the below code isn't necessary in example code.

Reply:
i think it illustrates that it actually works. I could have called CreateKey and provided a password and a pcr selection and subsequently called Unseal and claimed that it sealed against the PCR when in fact it would have not.

Reply:
I guess my question is: is this an example of general TPM functionality (i.e. spec conformance by the device) or an example of using the library?

Reply:
It's an example of using the library. Even so, it is nice to demonstrate concretely what the usefulness of the operation of interest is.